Add wifi density feature

elohanlon opened the pull request.

cgalarza commented on app/helpers/wifi_density_helper.rb:
```diff
@@ -0,0 +1,94 @@
+module WifiDensityHelper
+  CACHE_DURATION = Rails.env.development? ? 10.seconds : 5.minutes
```
Maybe this should be something that is configurable instead of baked into the code for one environment.

barmintor replied:
Seems like something we could probably add to the app config for density.

cgalarza commented on app/helpers/wifi_density_helper.rb:
```diff
@@ -0,0 +1,94 @@
+        aggregated_locations.merge!(recursively_collect_children(location_data))
+      end
+
+      # Sum client_count totals from all aggregated locations
```
I believe you should be able to use [Enumerable#reduce](https://ruby-doc.org/core-2.6/Enumerable.html#method-i-reduce) here.

cgalarza commented on app/helpers/wifi_density_helper.rb:
```diff
@@ -0,0 +1,94 @@
+    else
+      JSON.parse(Net::HTTP.get(URI(root_url + "wifi-density.json")))
+    end
+  end
```
I wonder if a configuration setting here would be better than just having the file paths listed in this code, but its a little tricky because one is a url request and one is not.

cgalarza replied:
I thought a little more about what I would do here. For the testing environment, I would use WebMock to mock the request for the wifi-density. That way you can also test that the response is being made the way you expected. Though this solution doesn't provide an alternative for development. Just a thought.

barmintor replied:
We should find a way to configure this.

cgalarza commented on config/initializers/wifi_density.rb:
```diff
@@ -0,0 +1,2 @@
+config_file_location = Rails.root.join('config/wifi_density.yml')
+WIFI_DENSITY = File.exists?(config_file_location) ? YAML.load_file(config_file_location) : {}
```
Could use `Rails.application.config_for(:wifi_density)`

barmintor replied:
This would also make it a little less annoying to override in a view test.

elohanlon commented:
> ```
> module DensityHelper
>   def wifi_density_percentage(location)
>     WifiDensity.new(location).percentage
>   end
> end
> ```

> ...

> Whether a class or module should be use could be debated. A module might be better since there's just one method you need and there isn't a need to store other variables on an object.

@cgalarza Good call about helper vs lib.  I moved all of the logic into lib, and ended up going for a module rather than a class since I'm not storing any state. After that, since the helper was just a one-line wrapper around a method call, I opted to just call the lib method directly from the view. I've been Googling around and didn't find any evidence of that being a bad practice, and gems often have code in lib that's meant to be used in views, so I'm hoping that seems reasonable. Also, I ended up stubbing a method instead of using webmock, so hopefully that seems okay too. What do you think?  I've pushed an updated commit with the new code. Thank you to you and @barmintor for the good feedback. I've incorporated your other suggestions into my commit as well. Let me know what you think!

We're still waiting for Candice to get back to us about expected "high" values for each location in wifi_density.yml before we can show density stats in production, so no rush.

elohanlon commented:
Also, @barmintor, I just documented the wifi_density.yml file with a template file -- wifi_density.template.yml.  When you said "spec", did you mean an rspec test or just any kind of specification document (like the template file)?

cgalarza commented on .gitignore:
```diff
@@ -24,5 +24,6 @@
 
 /config/database.yml
 /config/secrets.yml
+/config/wifi_density.yml
```
Does this need to be a secret? Or can we commit it to the repo?

elohanlon replied:
It's not a secret, but it may vary from environment to environment as we test new configurations based on uncertain wifi data "high" values.

cgalarza commented on app/views/locations/index.html.erb:
```diff
@@ -28,8 +28,14 @@
             <%= status %>
           <% end %>
         </span>
+        <%- if (status != 'CLOSED' || Rails.env.development?) && (density_percentage = Hours::WifiDensity.percentage_for(location)) %>
```
👍

cgalarza commented on spec/hours/wifi_density_spec.rb:
```diff
@@ -0,0 +1,99 @@
+  let(:sample_hierarchical_wifi_density_data) { JSON.parse file_fixture("sample-hierarchical-wifi-density-data.json").read }
+
+  before do
+    allow(described_class).to receive(:fetch_raw_wifi_density_data).and_return(sample_wifi_density_response)
```
I think this is totally fine. 👍
